### PR TITLE
Adding convective momentum transport to MSKF

### PR DESCRIPTION
**Pull request by alapaty-epa**
TYPE:  enhancement

KEYWORDS: Convective Momentum Transport, MSKF

SOURCE: Kiran Alapaty (US EPA)

DESCRIPTION OF CHANGES: Like many regional CP schemes, the MSKF lacked the inclusion of convective momentum transport (CMT) associated with cumulus convection. The CMT formulation developed by Zhang and McFarlane (JGR, 100, 1995) was implemented into the MSKF scheme to include this missing process that improves speed of propagation of storms and squall lines and associated changes in horizontal winds components.  

LIST OF MODIFIED FILES: 
M       phys/module_cu_mskf.F
M       phys/module_cumulus_driver.F
M       phys/module_physics_addtendc.F
M       phys/module_physics_init.F
M       dyn_em/module_first_rk_step_part2.F

TESTS CONDUCTED:  Two WRF simulations were performed for 2-days to study impacts of the CMT physics on convective systems: these are (1) without and (2) with CMT physics using the MSKF.  See the attached WORD document on the test cases and results from my simulations. 
[forGitHub.docx](https://github.com/wrf-model/WRF/files/3975740/forGitHub.docx)

The code has passed regression test using the classroom computer, and it is bit-for-bit with 1 vs multiple processor runs when this option is turned on. The jenkins test is also ok.

The figure below shows surface rainfall from two 18 h forecast at 30 km: CMT-off on the left and CMT-on on the right, convective rainfall on the top, and microphysics rainfall on the bottom. 

cmt-pair

RELEASE NOTE:  
Added Zhang and McFarlane (JGR, 100, 1995) convective momentum transport (CMT) to the MSKF scheme (improves speed of propagation of storms and squall lines and associated changes in horizontal winds components). The option is controlled by cmt_opt_flag inside module_cu_mskf.F, and turned on.

### CODE REVIEW

**Comment by davegill** [comment]
@alapaty-epa 
This PR needs to be re-based to the correct branch. Likely what is intended is `develop`, not `master`.

**Comment by davegill** [comment]
@alapaty-epa 
Kiran,
I would suggest grabbing a couple of the sentences from the description and testing, and then adding that as a release note. If people are using this scheme, they will want to know that this modification impacts them.

File: phys/module_cumulus_driver.F [hunk 1]
```diff
@@ -1057,9 +1057,10 @@ SUBROUTINE cumulus_driver(grid                                     &
                ,ZOL=zol,WSTAR=wstar,UST=ust,PBLH=pblh           & !ckay
                ,AEROCU=aerocu,NO_SRC_TYPES_CU=no_src_types_cu   &
                ,AERCU_FCT=aercu_fct,AERCU_OPT=aercu_opt        &
-               ,EFCS=EFCS,EFIS=EFIS,EFSS=EFSS)
-#endif
-
+               ,EFCS=EFCS,EFIS=EFIS,EFSS=EFSS                   &
+               ,RUCUTEN=RUCUTEN,RVCUTEN=RVCUTEN,XLAND=XLAND)       !JTR
+#endif  
+           print *,'here'
```
Comment (davegill) [inline phys/module_cumulus_driver.F]: @alapaty-epa 
Kiran,
Zap this debug statement

**Comment by alapaty-epa** [comment]
Thanks Dave ; will take it soon.

--
Kiran Alapaty  | MD E205-09  | 919 541 3130

**Comment by alapaty-epa** [comment]
Will do Dave soon

--
Kiran Alapaty  | MD E205-09  | 919 541 3130

**Comment by weiwangncar** [comment]
@alapaty-epa It will be nice to add some graphics to this PR so that we can see what kind of impact this change has in your test case. Things like track and intensity improvement, and in non-tropical storm runs, impact on precipitation, would all be interesting to see.

**Comment by alapaty-epa** [comment]
Thank you for your kind suggestion. Will prepare a couple to few graphs and will add them in the next week.

--
Kiran Alapaty  | MD E205-09  | 919 541 3130

**Comment by alapaty-epa** [comment]
dear all, as requested, i have done some analysis of simulations that uses new convective momentum transfer (CMT) physics to study CMT impacts on convective systems. Attached WORD doc contains all needed details and results. Hope this will be good enough for your consideration. Thanks for your patience.

**Comment by alapaty-epa** [comment]
All, note that in the attached write up in the line # 44, words "little but"  should be read as "little bit". 
it was a typo and changes meaning drastically :-(

**Comment by alapaty-epa** [comment]
please see the attached WORD document on my 90 day simulation (JJA 2006) for BASE and CMT cases using 12 km grids. A paper will be written soon:  "Introducing CMT physics to the regional climate modeling: Impacts on convection timing and location errors" .  I hope that these tests are good enough for your evaluation and consideration. 
[jja2006ForGitHub.docx](https://github.com/wrf-model/WRF/files/3989087/jja2006ForGitHub.docx)

File: phys/module_cumulus_driver.F (hunk 1 above)
Comment (weiwangncar) [reply phys/module_cumulus_driver.F]: @davegill This is done.

**Comment by weiwangncar** [comment]
@alapaty-epa How do you explain the relatively larger reduction of RAINC over the ocean in your JJA means?

**Comment by alapaty-epa** [comment]
Thank you for asking this important question.

CMT can enhance or decrease precipitation, as can be seen over the land and ocean regions. However, the feedbacks are stronger over the ocean due to strong feedbacks between CMT changes and dynamics. As documented in the literature, convection–evaporation–wind feedback, and condensation heating–circulation feedback,  in the climate system can be triggered by the CMT forcing. When wind fields are perturbed by the CMT forcing, the adjustment of dynamical advection results in the modification of temperature fields and hence convection ( and convective heating), which in turn can affect the temperature.
One need to dive into the circulation pattern changes (CMT minus Base), wind shear changes, temperature changes associated with two simulations to confirm the above mentioned feedbacks.  Over the ocean, organized convection during JJA seemed to be a predominant feature with larger precipitation (than over the land areas), thus the impacts of CMT are dramatic in the climate (JJA) simulation. Also, our experiences in the past was that over the atlantic ocean shown in the results experienced too much precipitation and thus the reduced precipitation due to CMT is a relief given the putative bias in the precip over this oceanic region.

More analysis along these lines is being conducted for use in a m/s.

--
Kiran Alapaty  | MD E205-09  | 919 541 3130

**Comment by weiwangncar** [comment]
@alapaty-epa @dudhia @davegill I did a test with this CMT (convective momentum transport) code using a 30 km CONUS case from 2016051200. The image below shows 18 h convective rain (top) and microphysics rain (bottom). The left column is with CMT-off, and the right, CMT-on. Differences clearly exist. 
cmt_on-off

**Comment by weiwangncar** [comment]
@alapaty-epa @dudhia @davegill However, in order to use CMT option (a switch inside the code), one must set no_src_types_cu = 10, alevsiz_cu = 30, even though aercu_opt can be set to 0 (meaning not using aerosols). That's the situation now. What do you think?

**Comment by alapaty-epa** [comment]
Thank you for sharing. it is nice that you used 30 km grids at which CP scheme is more active than the MP scheme and is a good case I think.  I am studying the timing of convection impacted by the CMT etc…  best, k

--
Kiran Alapaty  | MD E205-09  | 919 541 3130

**Comment by alapaty-epa** [comment]
I agree. In the default/release namelist.input file it helps to stick in these parameters for now and go from there. I really don’t know why this is the case but I am leaving it to your expertise...

--
Kiran Alapaty  | MD E205-09  | 919 541 3130

**Comment by weiwangncar** [comment]
@alapaty-epa However this trick doesn't work with gfortran... This is a problem...

**Comment by davegill** [comment]
@alapaty-epa @weiwangncar 
>In the default/release namelist.input file it helps to stick in these parameters for now and go from there. I really don’t know why this is the case but I am leaving it to your expertise

Kiran,
What does this mean?

**Comment by weiwangncar** [comment]
@davegill For ifort, the code won't run unless certain namelist is set (no_src_types_cu = 10, alevsiz_cu = 30). For gfortran, the code won't run even with these namelists in. I started to debug last week, and have not finished.

@alapaty-epa So far I've found, using 'configure -D' with gfortran compiler, 1) before you can mskf_cmt, msg1 needs to be initialized. What is the value to initialize this variable? 2) In mskf_cmt, do loop 500, there seems to be out of bound data access for J = ILEV. Can you comment please?

**Comment by alapaty-epa** [comment]
Sorry – one mistake inadvertently I did was building the CMT codes on top of aerosol codes in the early stage  BUT then, I quickly backed out and separated CMT code in the early stages of CMT implementation. This way, CMT will be independent of aerosol interactions. So, in that separation process,  I forgot to initialize the MSG1 to zero  - this flag is to skip any vertical levels to exclude impacts of the CMT - an option to study the impact of CMT on PBL vs free atmosphere – just for diagnostic purposes.

To cut short, please add line  MSG1 = 0  before the call mskf_cmt  and I am sorry to miss initializing it  ☹

Now, issue with the do loop 500:  I think it is arising due to messed up MSG1 variable.

Interestingly, neither my 6 day forecasts nor JJA seasonal run failed – plus CMT results were not strange as compared to the control run where CMT is not used.

Hopefully this will solve the mystery…
--
Kiran Alapaty  | MD E205-09  | 919 541 3130

**Comment by alapaty-epa** [comment]
Like Wei said, the CMT code fails when I delete lines: number of aerosol types and CESM aerosol vertical levels in the namelist.input file.

I think MSG1 is the culprit since MSG1 is also used in the aerosol codes – so, by defining/adding the above two parameters in the namelist.input file, the MSG1 gets initialized to zero in the aerosol portion of the code and thus CMT code runs fine.

This is my logic here but …

--
Kiran Alapaty  | MD E205-09  | 919 541 3130

**Comment by alapaty-epa** [comment]
I quickly rain the WRF again without initializing MSG1 and print statements show surprisingly  MSG1 is assigned a zero value – good thing.  However, I added a new line (MSG1=0) to my CMT code

DSUBCLD   567.253417968750
JBB          25
JDD          24
JTT          23
msg1           0
DT   60.0000000000000
grav   9.80000019073486
cpin   1004.50000000000
rdry   287.000000000000
KTE          35
IL1G           1
IL2G           1
ILG           1

--
Kiran Alapaty  | MD E205-09  | 919 541 3130

**Comment by weiwangncar** [comment]
@alapaty-epa What about this loop:
               DO KQ=KTS,KTE
                 JK = KTE-KQ+1
                 DUDTnew(1,JK) = 0.0
                 DVDTnew(1,JK) = 0.0
                 DPDX(1,JK) = 0.0
                 DPDY(1,JK) = 0.0

                 zf_wrf(KQ) = zf_wrf(KQ-1)+DZQ(KQ)
                 zfu(1,JK) = zf_wrf(KQ)

When KQ=KTS=1, what do you expect the value of zf_wrf to be at (kq-1)?

**Comment by alapaty-epa** [comment]
initialization of zf_wrf was done just below the cmt_opt_flag – pls see below lines:

        IF(cmt_opt_flag) THEN
              zf_wrf(0) = 0.0  ! ground
            ….

--
Kiran Alapaty  | MD E205-09  | 919 541 3130

**Comment by weiwangncar** [comment]
@alapaty-epa OK, thanks. I pointed out this line because after 2-h integration in debug mode, the code stopped with a 'Floating-point exception - erroneous arithmetic operation' error, pointing to these lines:

               DO KQ=KTS,KTE
                  stat_energy(KQ) = CP*(T0(KQ)*(1.0+0.622*Q0(KQ))) + G*Zf_wrf(KQ)
               ENDDO

**Comment by alapaty-epa** [comment]
Since there is no division here in that line – I am not sure why it is blowing up…

I again ran the code for 3 months and it did not crash at all when using the ifort

Thx.
--
Kiran Alapaty  | MD E205-09  | 919 541 3130

**Comment by weiwangncar** [comment]
@alapaty-epa @davegill Good news that when MSG1 is initialized, the ifort-compiled code on Cheyenne runs now without the need to add those aerosol namelist. I will update the code, and run another reg test today.

**Comment by alapaty-epa** [comment]
This is very good news. Thanks Wei. Hopefully code will work with the gfortran also

--
Kiran Alapaty  | MD E205-09  | 919 541 3130

**Comment by weiwangncar** [comment]
@alapaty-epa Do you have access to a gfortran compiler? If you do, can you try to compile and run the CMT code using that compiler? Using gfortran 7.3 on our local computer, and 8.3 on Cheyenne, the model fails within one hour. The code is causing MU to become NaN within one model time step.

The regression tests conducted on our classroom computer have passed when CMT physics is on. Note that this is a 10 model time step test only. And it uses gfortran 5.2.

**Comment by weiwangncar** [comment]
@alapaty-epa Speaking too soon: model output files from serial compilation versus MPI compilation are different - indicating there is still code issues with CMT physics on.

**Comment by alapaty-epa** [comment]
Thank you for letting me know.  MU is not required for the CMT codes and thus it seems  that here you are activating aerosol codes since MU is an array in the double moment convective microphysics of the MSKF aerosol codes and surely not required for CMT.
Please check and if so,  for now turn off aerosol physics.

Switching gears…
To the best of my knowledge, none in our entire group uses gfortran nor aware of such compiler availability. But,  I am contacting our system admins to see whether EPA can acquire it and install it on our computers OR to see if there is any other dinky computer that has gfortran.

Else I will see if anyone at NCSU can help me testing the codes using gfortran.

If you have any thoughts for me  pls let me know…
--
Kiran Alapaty  | MD E205-09  | 919 541 3130

**Comment by alapaty-epa** [comment]
Early this week, i ran the WRF using CMT codes for 3 months on our computers nothing looked strange in the results. So, moving away from the runs.

Now checking the code visually to find any potential culprits here in the CMT codes and will get back to you ASAP…

--
Kiran Alapaty  | MD E205-09  | 919 541 3130

**Comment by weiwangncar** [comment]
@alapaty-epa I am not using aerosol code from MSKF. The MU array I'm referring to is the dry column pressure of the model. This field will have trouble if the physics are not providing the right tendencies, for example.

**Comment by weiwangncar** [comment]
@alapaty-epa Found another undefined array, zf_wrf, which is not defined when stat_energy is computed. This array is defined only if the aerosol code is active. This change now allows me to integrate the model with CMT physics on using gfortran.

**Comment by davegill** [comment]
@dudhia 
Jimy,
A review please

**Comment by weiwangncar** [comment]
The code is now ok to merge. It has passed jenkins test, and when turned on, passed regtest in the classroom computer. I left the switch to use the convective momentum transport on. We can decide if we should change it.

**Comment by alapaty-epa** [comment]
Wei WHEN you get a chance – can you let me know what were those fixes you made CMT codes work in your tests. Though original codes are running fine on our machines but I want to make sure to use right results for a future paper by using your fixes to maintain consistency. Thx. k

--
Kiran Alapaty  | MD E205-09  | 919 541 3130

**Comment by weiwangncar** [comment]
@alapaty-epa Click on 'Files Changed' tab just above the PR, and you can see the files modified. The first shown there is the file I modified. The change is to call halo_em_phys_cu for cumulus schemes that produces momentum tendencies. This is because the momentum tendencies need to be averaged back to the staggered u, v points, hence needing the neighboring points.

**Comment by weiwangncar** [comment]
@alapaty-epa It would be nice if you can repeat some of your tests to see the impact.

**Comment by alapaty-epa** [comment]
Of course - Will do it this week and get back to you all..

--
Kiran Alapaty  | MD E205-09  | 919 541 3130

**Comment by weiwangncar** [comment]
Here is a repeat of my 30 km tests. Again, 18 h accumulated convective rainfall on the top, microphysics rainfall at the bottom, CMT off on the left, and CMT on at right. Largely similar to what they were before.

cmt-pair

**Comment by alapaty-epa** [comment]
Wei we have completed new runs using fixes you made. in general, there is an overall improvement in the squall line simulation. please see the attached powerpoint slides. 
[forGitHub-new.pptx](https://github.com/wrf-model/WRF/files/4383220/forGitHub-new.pptx)

**Comment by weiwangncar** [comment]
@alapaty-epa Thanks for posting the results. If you like, you could update the doc file you sent to me earlier to describe what CMT does, which we could point to in our release notes.

**Comment by alapaty-epa** [comment]
Sure no problem. I will update the write up with updated figures and post it at the github in a day or two.

--
Kiran Alapaty  | MD E205-09  | 919 541 3130

**Comment by alapaty-epa** [comment]
Since new simulations were performed after an update to the CMT code, attached document has updated results. The new results are much better after the fix. Thanks to Wei for her valuable contributions finding a couple of issues and making code functional on other platforms. 
[forGitHub-new.docx](https://github.com/wrf-model/WRF/files/4395004/forGitHub-new.docx)

**Pull request merged**